lib/openshift/cluster_version.rb

```diff
@@ -1,19 +1,70 @@
 module BushSlicer
   class ClusterVersion < ClusterResource
     RESOURCE = "clusterversions"
-    def channel(user: nil, cached: true, quiet: false)
+    def channel(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('spec', 'channel')
     end
 
-    def upstream(user: nil, cached: true, quiet: false)
+    def upstream(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('spec', 'upstream')
     end
 
-    def version(user: nil, cached: true, quiet: false)
+    def version(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('status', 'desired', 'version')
     end
+
+    def history(user: nil, cached: false, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'history')
+    end
+
+    # return a list of history matching the state we want to filter on
+    # query_hash is the parameter we want to filter on,
+    def history_matching(query_hash: nil , user: nil, cached: true, quiet: false)
+      key = query_hash.keys.first
+      value = query_hash[key]
+      self.history.select { |h| h[key] == value }
+    end
+
+    def image(user: nil, cached: true, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'desired', 'image')
+    end
+
+    # get the last entry of the spec.history which should be where the original
+    def image_base_url(user: nil, cached: true, quiet: false)
+      self.history.last['image'].split('@sha256:')[0]
+    end
+
+    def upgrade_completed?(target_version:, user:nil, cached: false, quiet: false)
+      res = self.history_matching(query_hash: {'version' => target_version})
+      if res.count > 0
+        logger.debug("MSG: #{res.first['message']}")
+        res.first['state'] == 'Completed'
+      else
+        false
+      end
+    end
+
+    def wait_for_upgrade_completion(target_version:, user:nil, cached: false, quiet: false, upgrade_timeout: 2*60*60)
```

Comment:
`s/upgrade_timeout/timeout/`, it is clear what is the timeout about in this context

Can we live without `target_version` parameter? Can we figure out by ourselves whether there is a pending update ATM or not?

Reply:
having this `target_version` parameter allows user to force upgrade to anything image he wants.

Reply:
The point is that we know whether an upgrade is in progress or not (per your other comment). Why do we need to ask user specify a version?

Reply:
IMO we wait for upgrade to finish, then we can check what version we are running.

Reply:
Having this parameter will allow the user to check if a `target_version` upgrade status from the history.  It's possible to have multiple upgrades

Reply:
With regards to `upgrade_timeout`.  I think I'll just take that parameter out since the wait method already has 2hrs as default.

Reply:
ack, rename or remove works for me

Reply:
other than that lgtm